database/src/main/kotlin/fi/vauhtijuoksu/vauhtijuoksuapi/database/impl/TimerDatabase.kt

```diff
@@ -0,0 +1,130 @@
+package fi.vauhtijuoksu.vauhtijuoksuapi.database.impl
+
+import fi.vauhtijuoksu.vauhtijuoksuapi.database.api.VauhtijuoksuDatabase
+import fi.vauhtijuoksu.vauhtijuoksuapi.database.configuration.DatabaseConfiguration
+import fi.vauhtijuoksu.vauhtijuoksuapi.database.models.TimerDbModel
+import fi.vauhtijuoksu.vauhtijuoksuapi.exceptions.ServerError
+import fi.vauhtijuoksu.vauhtijuoksuapi.models.Timer
+import io.vertx.core.Future
+import io.vertx.sqlclient.Row
+import io.vertx.sqlclient.RowSet
+import io.vertx.sqlclient.SqlClient
+import io.vertx.sqlclient.templates.SqlTemplate
+import mu.KotlinLogging
+import java.text.SimpleDateFormat
+import java.time.Instant
+import java.time.format.DateTimeFormatter
+import java.util.Date
+import java.util.UUID
+import javax.inject.Inject
+
+class TimerDatabase
+@Inject constructor(
+    private val client: SqlClient,
+    configuration: DatabaseConfiguration
+) : AbstractModelDatabase<Timer, TimerDbModel>(
+    client,
+    configuration,
+    "timers",
+    null,
+    { TimerDbModel -> TimerDbModel.toTimer() }
+),
+    VauhtijuoksuDatabase<Timer> {
+    private val logger = KotlinLogging.logger {}
+    val df = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ")
+
+    override fun <I, R> mapToFunction(template: SqlTemplate<I, R>): SqlTemplate<I, RowSet<TimerDbModel>> {
+        return template.mapTo(TimerDbModel::class.java)
+    }
+
+    @Suppress("MaxLineLength") // SQL is prettier without too many splits
+    override fun add(record: Timer): Future<Timer> {
+        return SqlTemplate.forUpdate(
+            client,
+            "INSERT INTO timers " +
+                "(start_time, end_time) VALUES " +
+                "(#{start_time}, #{end_time}) " +
+                "RETURNING *"
+        )
+            .mapFrom(TimerDbModel::class.java)
+            .mapTo(TimerDbModel::class.java)
+            .execute(TimerDbModel.fromTimer(record))
+            .recover {
+                throw ServerError("Failed to insert $record because of ${it.message}")
+            }
+            .map {
+                logger.debug { "Inserted timer $record" }
+                return@map it.iterator().next().toTimer()
+            }
+    }
+
+    @Suppress("MaxLineLength") // SQL is prettier without too many splits
+    override fun addMany(records: List<Timer>): Future<List<Timer>?> {
+        fun insertStatement(data: List<Timer>): String {
+            fun valuesStringForTimer(timer: Timer): String {
+                val startTime = if (timer.startTime != null) "'${df.format(timer.startTime)}'" else null
+                val endTime = if (timer.endTime != null) "'${df.format(timer.endTime)}'" else null
+                @Suppress("MaxLineLength")
+                return "('${timer.id}', $startTime, $endTime)"
+            }
+
+            var statement = "INSERT INTO timers VALUES "
+            for (timer in data) {
+                statement += "${valuesStringForTimer(timer)},"
```

Comment:
Building queries by appending Strings is a fine way to get SQL injections

Reply:
Probably not possible in this case, but I'd still rather not see code like this on production side

Reply:
To be honest, did not really consider that. Had a lot of issues finding a way to add many and this one was first to work but I admit it is not pretty. Main issue i think is mapping the list to query. For what is worth, I don't think this causes much security issue since it is still operating on Timer objects that only has dates and uuid. Of course I'm no cyber security expert. If you want some other implementation, I'm going to need a suggestion.

Reply:
> To be honest, did not really consider that. Had a lot of issues finding a way to add many and this one was first to work but I admit it is not pretty. Main issue i think is mapping the list to query. For what is worth, I don't think this causes much security issue since it is still operating on Timer objects that only has dates and uuid. Of course I'm no cyber security expert. If you want some other implementation, I'm going to need a suggestion.
How about using a separate table (that sounds like the best solution), but modifying the db client to fetch from both and constructing the object? You couldn't anymore use the abstract client, but it might make the code cleaner.